### ISSUE TITLE
`PyUnicode_Type` needs vectorcall.

### ISSUE THREAD

**Issue by markshannon** [python/cpython#117709]
Calls to `str()` must use the old `tp_call` protocol instead of `tp_vectorcall`, as `PyUnicode_Type` does not implement vectorcall.
All other common types, including `list`, `dict`, `tuple` implement vectorcall. For some reason `str` got overlooked.

### Linked PRs
* gh-117725
* gh-117746

**Comment by erlend-aasland** [comment]
> For some reason str got overlooked.

Perhaps since all three parameters can be passed as both keyword and positional arguments. That's a lot of combinations to spell out.

**Comment by erlend-aasland** [comment]
I made a PoC branch just for fun: https://github.com/erlend-aasland/cpython/pull/new/perf/unicode-vectorcall

Added paths for the cases where all args are positional, in additional to two positional and one keyword arg, and one positional and `encoding` & `errors` as keyword args in _that_ order only. Fallback to `tp_call` (build dict, etc.) for all other cases.

**Comment by erlend-aasland** [comment]
Some crude `timeit` benchmarks:

<details>
<summary>This PR</summary>

```console
$ ./python.exe -m timeit "str(1)"
5000000 loops, best of 5: 58.6 nsec per loop
$ ./python.exe -m timeit "str(b'a', 'latin1')"
10000000 loops, best of 5: 39.1 nsec per loop
$ ./python.exe -m timeit "str(b'a', 'latin1', errors='strict')"
5000000 loops, best of 5: 50 nsec per loop

# fallback to tp_call
$ ./python.exe -m timeit "str(object=b'a', encoding='latin1', errors='strict')"
2000000 loops, best of 5: 185 nsec per loop
```

</details>

<details>
<summary>`main`</summary>

```console
$ ./python.exe -m timeit "str(1)"
5000000 loops, best of 5: 56.7 nsec per loop
$ ./python.exe -m timeit "str(b'a', 'latin1')"
5000000 loops, best of 5: 75.4 nsec per loop
$ ./python.exe -m timeit "str(b'a', 'latin1', errors='strict')"
2000000 loops, best of 5: 145 nsec per loop
$ ./python.exe -m timeit "str(object=b'a', encoding='latin1', errors='strict')"
2000000 loops, best of 5: 182 nsec per loop
```

</details>

So there is a nice speedup for the fast paths that were added, no change for the single param case, and not a lot of overhead for the fallback cases. I guess the question now is how many cases we want to add; IOW, how much complexity do we allow.

**Comment by erlend-aasland** [comment]
Resolved with #117746 which speeds up `str()` calls with positional-only arguments.

**Comment by erlend-aasland** [comment]
We should consider "vectorcallising" `zip()` also.
